lib/invoice_generator.rb

```diff
@@ -40,13 +40,41 @@ def run
           end
 
           project_content[:resources].push(resource_content)
-          project_content[:cost] += resource_content[:cost]
+          project_content[:subtotal] += resource_content[:cost]
         end
 
+        # We first apply discounts then credits, this is more beneficial for users as it
+        # would be possible to cover total cost with fewer credits.
+        project_content[:cost] = project_content[:subtotal]
+        if project_records.first[:discount] > 0
```

Comment:
I was confused at the first sight to see `project_records.first` everywhere in this method. After further look, I realised it's about the group_by return type and you want to get the array of hashes into a shape of `|id, rest_of_the_content|`. For that, you can do something like;
```
arr.map! { |h| [h.delete("id"), h] }
```
This should give you the same functionality in-place and eliminate the need of group_by if I'm not mistaken.